### PR TITLE
Allow cops to restrict callbacks `on_send` to specific method names only

### PR DESCRIPTION
**Pull request by marcandre**
Many of the recent optimizations for Cops seem to be to modify `on_send` to check early for the `node.method_name` against a specific list of names.

This PR makes this pattern builtin so it is easy to do (just define `RESTRICT_ON_SEND = %i[foo bar baz]`) *and* makes it even more efficient.

Let's imagine 100 cops defining `on_send` that could have such a restricted method list. On a run over 10k `SendNode`, if these cops are optimized with the first thing in `on_send` being `return if SOME_LIST.include?(node.method_name)`, there would still be 100*10k calls to `on_send`, and 100*10k calls to `Set#include?` (assuming that `SOME_LIST` is a `Set`, as it should). With this PR, if the Cops are optimized with the right constant, then there would be only 10k hash lookups.

cc/ @fatkodima

### CODE REVIEW

File: lib/rubocop/cop/commissioner.rb [hunk 1]
```diff
@@ -7,6 +7,9 @@ module Cop
     class Commissioner
       include RuboCop::AST::Traversal
 
+      RESTRICTED_NODE_TYPES = %i[on_send on_csend].freeze
```
Comment (fatkodima) [inline lib/rubocop/cop/commissioner.rb:10]: Great work 👍 

This constant probably should be named differently, something like `RESTRICTED_CALLBACK_NAMES`.

Comment (marcandre) [reply lib/rubocop/cop/commissioner.rb:10]: Indeed 😄

**Comment by marcandre** [comment]
ping @bbatsov

**Comment by marcandre** [comment]
ping @bbatsov Let me know if my ping frequency is too high too 😅

**Comment by bbatsov** [comment]
The frequency is fine, but I couldn't catch up with all the open PRs. I like the proposed change, but I think we need to document it better if we expect someone to actually use it.

**Comment by marcandre** [comment]
Cool. I'll document this and then merge. Thanks!

**Comment by tas50** [comment]
@marcandre I reworked the 200+ cops in Cookstyle to use this new functionality where we could. Since we're autocorrecting a DSL there's a lot of times we're just looking for a deprecated method call that needs to be replaced with a new one. Lots and lots of simple on_sends. My testing shows this sped things up about 16% against a large set of cookbooks. Nice work.

**Comment by marcandre** [comment]
16% is awesome, especially given the minimal trouble it is to specify it! 🎉
Thank @fatkodima who started using this technique manually 😄

**Pull request merged**